ilalink/ilalink.1
.\" Copyright (c) 2001, 2002, 2003 Southern Storm Software, Pty Ltd.
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.TH ilalink 1 "14 February 2003" "Southern Storm Software" "Portable.NET Development Tools"
.SH NAME
ilalink \- the Intermediate Language (IL) assembly linker
.SH SYNOPSIS
.ll +8
.B ilalink
[ options ]
.I input
\&...
.SH DESCRIPTION
.B Ilalink
links a number of IL object files (.obj) together, resolves external
references to IL libraries, and creates a final IL executable (.exe)
or IL dynamic link library (.dll).
.SH OPTIONS
.TP
.B \-o FILE, \-\-output FILE
Specify the name of the file to write the linker output to.  The
default is "a.out.exe", "a.out.dll", or "a.out.obj", depending
on the value of the
.B \-\-format
option.
.TP
.B \-j, \-\-format obj
Write a ".obj" file as output, instead of an executable or dynamic
link library.  This is normally not very useful.
.TP
.B \-e, \-\-format exe
Write an executable file as output, instead of a ".obj" file or dynamic
link library.  This is the default value.
.TP
.B \-d, \-\-format dll
Write a dynamic link library file as output, instead of a ".obj" file or
executable.
.TP
.B \-l NAME, \-\-library NAME
Link against the library called "NAME".  The linker will search all of
the library directories for a file called "NAME.dll".  If \fB\-\-no\-stdlib\fR
is not specified, then the standard library "mscorlib" will be automatically
added to the list of libraries.  The name of the standard library can be
changed using the \fB\-fstdlib\-name\fR option.
.TP
.B \-L PATH, \-\-library\-dir PATH
Specify a directory to be searched for libraries.  The option `\fB\-L .\fR'
must be explicitly supplied to search the current directory.  If
\fB\-\-no\-stdlib\fR is not specified, then a standard set of builtin paths
will be added to the end of the list specified by the `\fB\-L\fR' options.
The set of builtin paths can be overridden by setting the "CSCC_LIB_PATH"
environment variable to an appropriate value.
.TP
.B \-n, \-\-no\-stdlib
Do not include the standard library or any of the builtin library directories.
.TP
.B \-S NAME, \-fstdlib\-name=NAME
Set the name of the standard library to "NAME".  The default is "mscorlib".
.TP
.B \-\-shared
Link the output as a shared image.  This is the default.
.TP
.B \-\-static
Link the output statically.  This option is not yet implemented.  It is
provided for compatibility with other tools.
.TP
.B \-a NAME, \-fassembly\-name=NAME
Set the name of the final assembly to "NAME".  The default is the name of
the output file without its directory path or extension.
.TP
.B \-A NN:NN:NN:NN, \-fassembly\-version=NN:NN:NN:NN
Set the final assembly version to "NN:NN:NN:NN".  The default is
"0:0:0:0", unless a non-zero version is supplied by one of the input files.
.TP
.B \-M NAME, \-fmodule\-name=NAME
Set the name of the final module to "NAME".  The default is the name of
the output file without its directory path.
.TP
.B \-E NAME, \-\-entry\-point NAME
Specify the name of the program's entry point.  This should only be used
if the output is an executable (.exe).  If this option is not present,
the linker will attempt to deduce which method is the entry point by
inspecting the names and signatures of methods within the input files.
The name is either a class name, or a class name plus method name.
The sequence "::" is used to separate the class and method names.
Examples are "MyClass", "MyNamespace.MyClass", and "MyNamespace.MyClass::Main".
.TP
.B \-r FILE, \-fresources=FILE
Specify the name of a resource file to be linked with the output assembly.
.TP
.B \-R, \-\-resources\-only
Create an output assembly that only contains resources.  This is useful
for building satellite resource assemblies for foreign languages.
.TP
.B \-p, \-fprivate\-resources
Mark the resources as private to the output assembly.
.TP
.B \-H ALG, \-fhash\-algorithm=ALG
Set the assembly hash algorithm to "ALG", which must be either
"SHA1" or "MD5".  The default is "SHA1".
.TP
.B \-V vN.N.NNNN, \-fmetadata\-version=vN.N.NNNN
Set the final metadata version to "vN.N.NNNN".  The default is
inferred from the metadata version on "mscorlib".
.TP
.B \-z, \-fminimize\-parameters
Minimize the size of the parameter definition table by discarding
parameters that don't strictly need to be present for correct execution.
.TP
.B \-3, \-m32bit\-only
Mark the final output assembly so that it can only be used on 32-bit
systems.  Use of this option is severely discouraged.
.TP
.B \-c, \-mcui\-subsystem
Link the final output assembly for the command-line subsystem
on Windows.  This is the default.
.TP
.B \-G, \-mgui\-subsystem
Link the final output assembly for the GUI subsystem on Windows.
.TP
.B \-\-help
Print a usage message for the \fBilalink\fR program.
.TP
.B \-v, \-\-version
Print the version of the \fBilalink\fR program.
.TP
.B \-\-
Marks the end of the command-line options, and the beginning of
the input filenames.  You may need to use this if your filename
begins with '-'.  e.g. "ilalink -- -input.obj".  This is not needed
if the input is stdin: "ilalink -" is perfectly valid.
.PP
The `\fB\-f\fR' and `\fB\-m\fR' options can be specified on the \fBcscc\fR(1)
command-line, and will be passed to the linker automatically by the compiler.
.SH "AUTHOR"
Written by Southern Storm Software, Pty Ltd.

http://www.southern-storm.com.au/
.SH "SEE ALSO"
ilasm(1), ildasm(1), cscc(1)
.SH "DIAGNOSTICS"
Exit status is 1 if an error occurred while processing the input.
Otherwise the exit status is 0.


NEWS

0.6.12 (16 January 2005)

	Runtime engine:

	* Beginning of PPC unroller implementation (Gopal V).

	C# Compiler:

	* Compatibility with /debug flags in csc (Gopal V).
	* Handling of method/property/etc declarations that hide base
	  class virtuals and access checks for same (Richard Baumann).

	Documentation:

	* Information on debugging the unroller with gdb (Gopal V).

	Platform Support:

	* Use a portable version of tparm to match Susv3 specs (Gopal V).
	* Case sensitivity during comparisons (Klaus Treichel).
	* EAX vs REG_EAX usage in interrupt_posix (Gopal V).
	* Fix cache flush logic for PPC (Gopal V).
	* Skip empty lines in "@file" response files (Stephen Compall).

0.6.10 (30 October 2004)

	Runtime engine:

	* Null parameter names when emitting parameters (Gopal V).
	* Various fixes to System.Reflection internalcalls (Gopal V).
	* Thread suspend/resume (Thong Nguyen).
	* Internalcalls for CustomAttributeBuilder (Rhys Weatherley).
	* Use "fprem", not "fprem1" for x86 IL remainder (Rhys Weatherley).
	* Handle vararg methods and delegates more centrally (Gopal V).
	* System.Decimal division bug (Klaus Treichel).
	* Do not inline loads from value types or crashes will occur (Gopal V).

	C# Compiler:

	* Work around some PPC64-related optimization problesm (Gopal V).
	* Nested class "famorassem" handling (Gopal V).
	* Fixes to enum initializers (Gopal V).

	BF Compiler:

	* Fix the behavior of BFOptimize() which currently destroys the
	  parse tree (Gopal V).

	Loader and Metadata:

	* Prevent double loading of assemblies by using the full path (Gopal V).
	* Use ILNativeUInt instead of "unsigned long" in il_types.h, which
	  fixes some problems with 64-bit platforms (Rhys Weatherley).
	* Properly import types before writing them out with SR.Emit (Gopal V).
	* Add a flag to enable ignoring load errors (Rhys Weatherley).

	Assembler:

	* Forward references for data labels (Gopal V).

	Platform Support:

	* Get user name via uid, rather than getlogin (Carl-Adam Brengesjo).
	* Enable threading for Solaris (Robin Kay).
	* Fix boot time calculation at the nanosecond level (Gopal V).
	* Bug in socket selection for infinite selects (Gopal V).
	* Alter gen_errno.sh to handle errno codes that are split across
	  multiple header files (Rhys Weatherley).
	* Response file support in cpp for systems with limited command
	  line lengths (Maciek Plewa).
	* Socket flags for Solaris (Adam J. Tybor).
	* Port socket code to hpux by including <sys/modem.h> (Gopal V).
	* Fix daylight savings for Win32 (Carl-Adam Brengesjo).
	* Libffi bug in x86-64, causing incorrect stack alignment (Gopal V).
	* Serial I/O routines for Win32 (Gopal V).
	* Timeouts in Unix version of serial reads (Adam Gibson).
	* Various profile fixes (Gopal V).
	* BeOS port (Tomas Wilhelmsson).
	* Cris port (Simon Posnjak).
	* Man page formatting problems (Peter Lund).

	Other:

	* Clean up Debian packaging (Russell Stuart).

0.6.8 (8 August 2004)

	Runtime engine:

	* Module.GetTypes and Assembly.GetExportedTypes internalcall (Gopal V).
	* Fixes to IL offset determination in the method cache (Gopal V).
	* "synchronized" method handling (Thong Nguyen).
	* Be careful not to notify finalizers during allocation (Thong Nguyen).
	* Inlining of Math functions (Thong Nguyen).
	* Inlining simple properties that get/set fields (Thong Nguyen).
	* Various fixes in the finalizer code (Thong Nguyen).
	* Thread aborting (Thong Nguyen).
	* Using signals to detect NULL references and the like (Thong Nguyen).
	* Thread execution contexts (Thong Nguyen).
	* Dns and Socket internalcalls (Gopal V).
	* GCAddrOfPinnedObject fixes (Llewellyn Pritchard).
	* Extended console support (Rhys Weatherley).

	C# Compiler:

	* Correct line numbers for argumentvar nodes (Gopal V).
	* Fixes to identifier resolution (Gopal V).
	* Coercion fixes in the arguments to user-defined operators (Gopal V).
	* Create nested classes in the proper order (Gopal V).
	* Null values and IntPtr (Gopal V).
	* Fixes for indexers on value types (Gopal V).
	* Assign add operations involving strings and objects (Gopal V).
	* Standalone assembly attributes (Rhys Weatherley).
	* Resolve classes before allocating heap objects (Gopal V).
	* Fix do ... while grammar to require a terminating ';' (Gopal V).
	* Add the "-fhidebyname" and "-fhidebysig" options (Rhys Weatherley).
	* Error message for private interface modifiers (Gopal V).
	* Check for event types before performing add/remove operations (Gopal V).

	C Compiler:

	* Fix ILNode_GetType for C strings (Gopal V).
	* Fix semantic analysis for member accesses (Gopal V).

	Loader and Metadata:

	* CompareUnicode not working correctly (Thong Nguyen).
	* ILClassImplements on a class should return true for itself (T. Nguyen).
	* Serialization of null strings in attribute blocks (Rhys Weatherley).
	* Recognise "OptionalAttribute" and "DefaultValueAttribute" (Rhys W).
	* Resolve classes in "TypeAssignCompatible" (Gopal V).
	* Recognise generic classes in files generated by .NET 2.0 (Rhys W).

	Linker:

	* Resource linking problems (Llewellyn Pritchard).
	* Obey "AssemblyVersionAttribute" if it is present (Rhys Weatherley).
	* "-fmetadata-version" option (Rhys Weatherley).
	* "-fculture" option (Rhys Weatherley).

	Platform Support:

	* Compilation fixes for 64-bit (Andrzej Dopierala).
	* Utilities for lists and queues (Thong Nguyen).
	* Various windows build fixes (Thong Nguyen).
	* ILGetSinceRebootTime for non-Windows platforms (Thong Nguyen).
	* Faster versions of volatile and interlocked functions (Thong Nguyen).
	* Daylight savings problem (David Logan).

ilalink/ilalink.c
/*
 * ilalink.c - Link IL images together to form an assembly.
 *
 * Copyright (C) 2001  Southern Storm Software, Pty Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <stdio.h>
#include "il_system.h"
#include "il_image.h"
#include "il_program.h"
#include "il_writer.h"
#include "il_linker.h"
#include "il_utils.h"

#ifdef	__cplusplus
extern	"C" {
#endif

/*
 * Table of command-line options.
 */
static ILCmdLineOption const options[] = {
	{"-o", 'o', 1, 0, 0},
	{"--output", 'o', 1,
		"--output file               or -o file",
		"Specify the output file to use.  The default name is\n"
		"based on the name of the first input file."},
	{"-e", 'e', 0,
		"--format exe                or -e",
		"Set the output format to executable."},
	{"-d", 'd', 0,
		"--format dll                or -d",
		"Set the output format to dynamic link library."},
	{"-j", 'j', 0,
		"--format obj                or -j",
		"Set the output format to object file."},
	{"--format", 'F', 1, 0, 0},
	{"-l", 'l', 1, 0, 0},
	{"--library", 'l', 1,
		"--library name              or -l name",
		"Link against the library `name'."},
	{"-L", 'L', 1, 0, 0},
	{"--library-dir", 'L', 1,
		"--library-dir name          or -L name",
		"Look for libraries in the directory `name'."},
	{"-n", 'n', 0,
		"--no-stdlib                 or -n",
		"Do not link against the standard library."},
	{"-S", 'S', 1,
		"-fstdlib-name=name          or -S name",
		"Specify the name of the standard library (default is `mscorlib')."},
	{"--shared", '1', 0,
		"--shared",
		"Link the output as a shared image (default)."},
	{"--static", '2', 0,
		"--static",
		"Link the output as a static image."},
	{"-a", 'a', 1,
		"-fassembly-name=name        or -a name",
		"Specify the name of the assembly to embed in the output."},
	{"-A", 'A', 1,
		"-fassembly-version=version  or -A version",
		"Specify the assembly version to embed in the output."},
	{"-M", 'M', 1,
		"-fmodule-name=name          or -M name",
		"Specify the name of the module to embed in the output."},
	{"-E", 'E', 1, 0, 0},
	{"--entry-point", 'E', 1,
		"--entry-point name          or -E name",
		"Specify the name of the entry point for the program."},
	{"-r", 'r', 1,
		"-fresources=file            or -r file",
		"Link the specified resource file with the output."},
	{"-R", 'R', 0, 0, 0},
	{"--resources-only", 'R', 0,
		"--resources-only            or -R",
		"Create an output that only contains resources."},
	{"-p", 'p', 0,
		"-fprivate-resources         or -p",
		"Mark the resources as private to the output assembly."},
	{"-H", 'H', 1,
		"-fhash-algorithm=name       or -H name",
		"Specify the algorithm to use to hash files (SHA1 or MD5)"},
	{"-3", '3', 0,
		"-m32bit-only                or -3",
		"The resulting output file can only be used on 32-bit systems."},
	{"-c", 'c', 0,
		"-mcui-subsystem             or -c",
		"Compile for the command-line subsystem (default)."},
	{"-G", 'G', 0,
		"-mgui-subsystem             or -G",
		"Compile for the GUI subsystem."},
	{"-f", 'f', 1, 0, 0},
	{"-m", 'm', 1, 0, 0},
	{"--version", 'v', 0,
		"--version                   or -v",
		"Print the version of the program"},
	{"--help", 'h', 0,
		"--help",
		"Print this help message."},
	{0, 0, 0, 0, 0}
};

static void usage(const char *progname);
static void version(void);
static void outOfMemory(void);
static int parseVersion(ILUInt16 *version, const char *str);
static int addLibrary(ILLinker *linker, const char *filename);
static int addResource(ILLinker *linker, const char *filename,
					   FILE *stream, int privateResources, int isStdin);
static int processFile(ILLinker *linker, const char *filename,
					   FILE *stream, int isStdin);

int main(int argc, char *argv[])
{
	char *progname = argv[0];
	int sawStdin;
	int state, opt;
	char *param;
	FILE *outfile;
	FILE *infile;
	int errors = 0;
	char *outputFile = NULL;
	int format = -1;
	int flags = 0;
	int resourcesOnly = 0;
	char *assemblyName = NULL;
	ILUInt16 assemblyVersion[4] = {0, 0, 0, 0};
	char *moduleName = NULL;
	char *entryPoint = NULL;
	int hashAlgorithm = IL_META_HASHALG_SHA1;
	int len;
	char **libraries;
	int numLibraries = 0;
	char **libraryDirs;
	int numLibraryDirs = 0;
	char *stdLibrary = "mscorlib";
	char **resources;
	int numResources = 0;
	int jvmMode = 0;
	int useStdlib = 1;
	int isStatic = 0;
	int privateResources = 0;
	int temp, temp2;
	ILLinker *linker;

	/* Allocate an array to hold the libraries to link against */
	libraries = (char **)ILCalloc(argc, sizeof(char *));
	if(!libraries)
	{
		outOfMemory();
	}

	/* Allocate an array to hold the directories to look for libraries in */
	libraryDirs = (char **)ILCalloc(argc, sizeof(char *));
	if(!libraryDirs)
	{
		outOfMemory();
	}

	/* Allocate an array to hold the resource files to link against */
	resources = (char **)ILCalloc(argc, sizeof(char **));
	if(!resources)
	{
		outOfMemory();
	}

	/* Parse the command-line arguments */
	state = 0;
	while((opt = ILCmdLineNextOption(&argc, &argv, &state,
									 options, &param)) != 0)
	{
		switch(opt)
		{
			case 'o':
			{
				outputFile = param;
			}
			break;

			case 'e':
			{
				format = IL_IMAGETYPE_EXE;
			}
			break;

			case 'd':
			{
				format = IL_IMAGETYPE_DLL;
			}
			break;

			case 'j':
			{
				format = IL_IMAGETYPE_OBJ;
			}
			break;

			case 'F':
			{
				if(!strcmp(param, "exe"))
				{
					format = IL_IMAGETYPE_EXE;
				}
				else if(!strcmp(param, "dll"))
				{
					format = IL_IMAGETYPE_DLL;
				}
				else if(!strcmp(param, "obj"))
				{
					format = IL_IMAGETYPE_OBJ;
				}
				else
				{
					usage(progname);
					return 1;
				}
			}
			break;

			case 'l':
			{
				libraries[numLibraries++] = param;
			}
			break;

			case 'L':
			{
				libraryDirs[numLibraryDirs++] = param;
			}
			break;

			case 'n':
			{
				useStdlib = 0;
			}
			break;

			case 'a':
			{
				assemblyName = param;
			}
			break;

			case 'A':
			{
				if(!parseVersion(assemblyVersion, param))
				{
					fprintf(stderr,
							"%s: `%s' is not a valid assembly version\n",
							progname, param);
					return 1;
				}
			}
			break;

			case 'M':
			{
				moduleName = param;
			}
			break;

			case 'E':
			{
				entryPoint = param;
			}
			break;

			case 'r':
			{
				resources[numResources++] = param;
			}
			break;

			case 'R':
			{
				resourcesOnly = 1;
			}
			break;

			case 'p':
			{
				privateResources = 1;
			}
			break;

			case 'S':
			{
				stdLibrary = param;
			}
			break;

			case '1':
			{
				isStatic = 0;
			}
			break;

			case '2':
			{
				isStatic = 1;
			}
			break;

			case 'H':
			{
			parseHashAlg:
				if(!ILStrICmp(param, "SHA1") || !ILStrICmp(param, "SHA-1"))
				{
					hashAlgorithm = IL_META_HASHALG_SHA1;
				}
				else if(!ILStrICmp(param, "MD5"))
				{
					hashAlgorithm = IL_META_HASHALG_MD5;
				}
				else
				{
					fprintf(stderr,
							"%s: `%s' is not a valid hash algorithm.\n",
							progname, param);
					fprintf(stderr, "Valid values are `SHA1' and `MD5'.\n");
					return 1;
				}
			}
			break;

			case 'f':
			{
				/* Parse a flag passed down from the compiler */
				if(!strncmp(param, "stdlib-name=", 12))
				{
					stdLibrary = param + 12;
				}
				else if(!strncmp(param, "assembly-name=", 14))
				{
					assemblyName = param + 14;
				}
				else if(!strncmp(param, "assembly-version=", 17))
				{
					if(!parseVersion(assemblyVersion, param + 17))
					{
						fprintf(stderr,
								"%s: `%s' is not a valid assembly version\n",
								progname, param);
						return 1;
					}
				}
				else if(!strncmp(param, "module-name=", 12))
				{
					moduleName = param + 12;
				}
				else if(!strncmp(param, "resources=", 10))
				{
					resources[numResources++] = param + 10;
				}
				else if(!strncmp(param, "hash-algorithm=", 15))
				{
					param += 15;
					goto parseHashAlg;
				}
				else if(!strcmp(param, "private-resources"))
				{
					privateResources = 1;
				}
				else
				{
					/* All other flags are ignored, because they may
					   be for other parts of the compiler chain */
				}
			}
			break;

			case 'm':
			{
				/* Parse a machine flag passed down from the compiler */
				if(!strcmp(param, "32bit-only"))
				{
					flags |= IL_WRITEFLAG_32BIT_ONLY;
				}
				else if(!strcmp(param, "cui-subsystem"))
				{
					flags &= ~IL_WRITEFLAG_SUBSYS_GUI;
				}
				else if(!strcmp(param, "gui-subsystem"))
				{
					flags |= IL_WRITEFLAG_SUBSYS_GUI;
				}
				else if(!strcmp(param, "jvm"))
				{
					jvmMode = 1;
				}
				else
				{
					/* All other flags are ignored, because they may
					   be for other parts of the compiler chain */
				}
			}
			break;

			case '3':
			{
				flags |= IL_WRITEFLAG_32BIT_ONLY;
			}
			break;

			case 'c':
			{
				flags &= ~IL_WRITEFLAG_SUBSYS_GUI;
			}
			break;

			case 'G':
			{
				flags |= IL_WRITEFLAG_SUBSYS_GUI;
			}
			break;

			case 'v':
			{
				version();
				return 0;
			}
			/* Not reached */

			default:
			{
				usage(progname);
				return 1;
			}
			/* Not reached */
		}
	}

	/* We need at least one input file argument */
	if(argc <= 1)
	{
		usage(progname);
		return 1;
	}

	/* Determine the default output format from the
	   extension on the output file */
	if(format == -1)
	{
		format = IL_IMAGETYPE_EXE;
		if(outputFile)
		{
			len = strlen(outputFile);
			while(len > 0 && outputFile[len - 1] != '.' &&
				  outputFile[len - 1] != '/' &&
				  outputFile[len - 1] != '\\')
			{
				--len;
			}
			if(len > 0 && outputFile[len - 1] == '.')
			{
				if(!ILStrICmp(outputFile + len, "dll"))
				{
					format = IL_IMAGETYPE_DLL;
				}
				else if(!ILStrICmp(outputFile + len, "obj"))
				{
					format = IL_IMAGETYPE_OBJ;
				}
			}
		}
	}

	/* Set the default output filename if necessary */
	if(!outputFile)
	{
		if(format == IL_IMAGETYPE_EXE)
		{
			outputFile = "a.out.exe";
		}
		else if(format == IL_IMAGETYPE_DLL)
		{
			outputFile = "a.out.dll";
		}
		else
		{
			outputFile = "a.out.obj";
		}
	}

	/* Determine the default module and assembly names */
	temp = strlen(outputFile);
	temp2 = -1;
	while(temp > 0 && outputFile[temp - 1] != '/' &&
		  outputFile[temp - 1] != '\\')
	{
		--temp;
		if(outputFile[temp] == '.' && temp2 == -1)
		{
			temp2 = temp;
		}
	}
	if(!moduleName)
	{
		moduleName = ILDupString(outputFile + temp);
	}
	if(!assemblyName)
	{
		assemblyName = ILDupString(outputFile + temp);
		if(temp2 != -1)
		{
			assemblyName[temp2 - temp] = '\0';
		}
	}

	/* Open the output file */
	if((outfile = fopen(outputFile, "wb")) == NULL)
	{
		/* Try again in case libc does not understand "wb" */
		if((outfile = fopen(outputFile, "w")) == NULL)
		{
			perror(outputFile);
			return 1;
		}
	}

	/* Create the linker context */
	linker = ILLinkerCreate(outfile, 1, format, flags);
	if(!linker)
	{
		fclose(outfile);
		ILDeleteFile(outputFile);
		outOfMemory();
	}

	/* Create the global module and assembly */
	if(!ILLinkerCreateModuleAndAssembly(linker, moduleName,
										assemblyName, assemblyVersion,
										hashAlgorithm))
	{
		ILLinkerDestroy(linker);
		fclose(outfile);
		ILDeleteFile(outputFile);
		outOfMemory();
	}

	/* Add the library directories to the linker context */
	for(temp = 0; temp < numLibraryDirs; ++temp)
	{
		errors |= !ILLinkerAddLibraryDir(linker, libraryDirs[temp]);
	}
	if(useStdlib)
	{
		if(!ILLinkerAddSystemDirs(linker))
		{
			errors = 1;
		}
	}

	/* Add the libraries to the linker context */
	for(temp = 0; temp < numLibraries; ++temp)
	{
		errors |= addLibrary(linker, libraries[temp]);
	}
	if(useStdlib)
	{
		errors |= addLibrary(linker, stdLibrary);
	}

	/* Process the input files that aren't libraries */
	sawStdin = 0;
	while(argc > 1)
	{
		if(!strcmp(argv[1], "-"))
		{
			/* Process the contents of stdin, but only once */
			if(!sawStdin)
			{
				if(resourcesOnly)
				{
					errors |= addResource(linker, "stdin", stdin,
										  privateResources, 1);
				}
				else
				{
					errors |= processFile(linker, "stdin", stdin, 1);
				}
				sawStdin = 1;
			}
		}
		else
		{
			/* Process the contents of a regular file */
			if((infile = fopen(argv[1], "rb")) == NULL)
			{
				/* Try again in case libc did not understand the 'b' */
				if((infile = fopen(argv[1], "r")) == NULL)
				{
					perror(argv[1]);
					errors = 1;
					++argv;
					--argc;
					continue;
				}
			}
			if(resourcesOnly)
			{
				errors |= addResource(linker, argv[1], infile,
									  privateResources, 0);
			}
			else
			{
				errors |= processFile(linker, argv[1], infile, 0);
			}
		}
		++argv;
		--argc;
	}

	/* Add the explicit resource files to the linker context */
	for(temp = 0; temp < numResources; ++temp)
	{
		if(!strcmp(resources[temp], "-"))
		{
			/* Process the contents of stdin, but only once */
			if(!sawStdin)
			{
				errors |= addResource(linker, "stdin", stdin,
									  privateResources, 1);
				sawStdin = 1;
			}
		}
		else
		{
			if((infile = fopen(resources[temp], "rb")) == NULL)
			{
				/* Try again with "r" in case libc does not understand "rb" */
				if((infile = fopen(resources[temp], "r")) == NULL)
				{
					perror(resources[temp]);
					errors = 1;
				}
			}
			if(infile != NULL)
			{
				errors |= addResource(linker, resources[temp], infile,
									  privateResources, 0);
			}
		}
	}

	/* Set the default entry point, or report an error about
	   incorrect entry point specifications */
	if(!resourcesOnly)
	{
		if(entryPoint)
		{
			if(!ILLinkerSetEntryPoint(linker, entryPoint))
			{
				fprintf(stderr, "%s: could not locate the entry point `%s'",
						outputFile, entryPoint);
				errors = 1;
			}
		}
		if(format == IL_IMAGETYPE_EXE)
		{
			if(!ILLinkerHasEntryPoint(linker))
			{
				fprintf(stderr, "%s: no entry point specified\n", outputFile);
				errors = 1;
			}
		}
		else if(format == IL_IMAGETYPE_DLL)
		{
			if(ILLinkerHasEntryPoint(linker))
			{
				fprintf(stderr, "%s: DLL's cannot have entry points\n",
						outputFile);
				errors = 1;
			}
		}
	}

	/* Destroy the linker context */
	len = ILLinkerDestroy(linker);
	fclose(outfile);
	if(!len)
	{
		ILDeleteFile(outputFile);
		errors = 1;
	}
	else if(len < 0)
	{
		ILDeleteFile(outputFile);
		outOfMemory();
	}

	/* Done */
	return errors;
}

static void usage(const char *progname)
{
	fprintf(stderr, "ILALINK " VERSION " - IL Assembly Link Utility\n");
	fprintf(stderr, "Copyright (c) 2001 Southern Storm Software, Pty Ltd.\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Usage: %s [options] input ...\n", progname);
	fprintf(stderr, "\n");
	ILCmdLineHelp(options);
}

static void version(void)
{

	printf("ILALINK " VERSION " - IL Assembly Link Utility\n");
	printf("Copyright (c) 2001 Southern Storm Software, Pty Ltd.\n");
	printf("\n");
	printf("ILALINK comes with ABSOLUTELY NO WARRANTY.  This is free software,\n");
	printf("and you are welcome to redistribute it under the terms of the\n");
	printf("GNU General Public License.  See the file COPYING for further details.\n");
	printf("\n");
	printf("Use the `--help' option to get help on the command-line options.\n");
}

static void outOfMemory(void)
{
	fputs("virtual memory exhausted\n", stderr);
	exit(1);
}

/*
 * Parse an assembly version string.
 */
static int parseVersion(ILUInt16 *version, const char *str)
{
	int posn;
	ILUInt32 value;
	for(posn = 0; posn < 4; ++posn)
	{
		if(*str < '0' || *str > '9')
		{
			return 0;
		}
		value = (ILUInt32)(*str++ - '0');
		while(*str >= '0' && *str <= '9')
		{
			value = value * ((ILUInt32)10) + (ILUInt32)(*str++ - '0');
			if(value >= (ILUInt32)0x10000)
			{
				return 0;
			}
		}
		version[posn] = (ILUInt16)value;
		if(posn == 3)
		{
			break;
		}
		if(*str != ':' && *str != '.')
		{
			return 0;
		}
		++str;
	}
	return (*str == '\0');
}

/*
 * Add a library to a linker context.  Returns non-zero on error.
 */
static int addLibrary(ILLinker *linker, const char *filename)
{
	FILE *file;
	ILContext *context;
	int loadError;
	ILImage *image;
	int errors = 0;
	char *newFilename;

	/* Resolve the library name */
	newFilename = ILLinkerResolveLibrary(linker, filename);
	if(!newFilename)
	{
		fprintf(stderr, "%s: library not found\n", filename);
		return 1;
	}

	/* Open the library file */
	if((file = fopen(newFilename, "rb")) == NULL)
	{
		/* Try again in case libc does not understand "rb" */
		if((file = fopen(newFilename, "r")) == NULL)
		{
			perror(newFilename);
			ILFree(newFilename);
			return 1;
		}
	}

	/* Load the library as an image */
	context = ILContextCreate();
	if(!context)
	{
		outOfMemory();
	}
	loadError = ILImageLoad(file, newFilename, context, &image,
							IL_LOADFLAG_FORCE_32BIT | IL_LOADFLAG_NO_RESOLVE);
	if(loadError)
	{
		fprintf(stderr, "%s: %s\n", newFilename, ILImageLoadError(loadError));
		ILContextDestroy(context);
		fclose(file);
		ILFree(newFilename);
		return 1;
	}
	fclose(file);

	/* Add the library image to the linker context */
	if(!ILLinkerAddLibrary(linker, image, newFilename))
	{
		errors = 1;
	}

	/* Clean up and exit */
	ILContextDestroy(context);
	ILFree(newFilename);
	return errors;
}

/*
 * Add a resource file to the final image.
 */
static int addResource(ILLinker *linker, const char *filename,
					   FILE *stream, int privateResources, int isStdin)
{
	int len;

	/* Determine the name of the resource from the filename */
	len = strlen(filename);
	while(len > 0 && filename[len - 1] != '/' &&
	      filename[len - 1] != '\\')
	{
		--len;
	}

	/* Add the resource to the linker image */
	if(!ILLinkerAddResource(linker, filename + len, privateResources, stream))
	{
		if(!isStdin)
		{
			fclose(stream);
		}
		return 1;
	}

	/* Done */
	if(!isStdin)
	{
		fclose(stream);
	}
	return 0;
}

/*
 * Process an image file to be linked into the final image.
 * Returns non-zero on error.
 */
static int processFile(ILLinker *linker, const char *filename,
					   FILE *stream, int isStdin)
{
	int errors = 0;
	ILContext *context;
	ILImage *image;
	int loadError;

	/* Attempt to load the image into memory */
	context = ILContextCreate();
	if(!context)
	{
		outOfMemory();
	}
	loadError = ILImageLoad(stream, filename, context, &image,
							IL_LOADFLAG_FORCE_32BIT | IL_LOADFLAG_NO_RESOLVE);
	if(!isStdin)
	{
		fclose(stream);
	}
	if(loadError != 0)
	{
		fprintf(stderr, "%s: %s\n", filename, ILImageLoadError(loadError));
		errors = 1;
	}
	else
	{
		/* Add the image to the linker context */
		if(!ILLinkerAddImage(linker, image, filename))
		{
			errors = 1;
		}
	}

	/* Clean up and exit */
	ILContextDestroy(context);
	return errors;
}

#ifdef	__cplusplus
};
#endif
